source/common/rate_limiter_impl.cc

```diff
@@ -52,6 +52,36 @@ void BurstingRateLimiter::releaseOne() {
   previously_releasing_ = absl::nullopt;
 }
 
+ScheduledStartingRateLimiter::ScheduledStartingRateLimiter(
+    RateLimiterPtr&& rate_limiter, Envoy::MonotonicTime scheduled_starting_time)
+    : ForwardingRateLimiterImpl(std::move(rate_limiter)),
+      scheduled_starting_time_(scheduled_starting_time) {
+  if (timeSource().monotonicTime() >= scheduled_starting_time_) {
+    throw NighthawkException("Scheduled starting time needs to be in the future");
+  }
+}
+
+bool ScheduledStartingRateLimiter::tryAcquireOne() {
+  if (timeSource().monotonicTime() < scheduled_starting_time_) {
+    aquisition_attempted_ = true;
+    return false;
+  }
+  // If we start forwarding right away on the first attempt that is remarkable, so leave a hint
+  // about this happening in the logs.
+  if (!aquisition_attempted_) {
+    aquisition_attempted_ = true;
+    ENVOY_LOG(warn, "First acquisition attempt was late");
+  }
+  return rate_limiter_->tryAcquireOne();
+}
+
+void ScheduledStartingRateLimiter::releaseOne() {
+  if (timeSource().monotonicTime() < scheduled_starting_time_) {
```

Comment:
I'm probably missing something, but couldn't a cancel operation (not yet implemented) result in releasing before the time arrives?  What are we trying to prevent by failing here?

Reply:
Right now, it is only valid to release what has been acquired. So when we hit this, it ought to be impossible to have a preceding successful acquisition because we're not due to start yet. The intent is to catch programming mistakes early when consuming this, but possibly this is overly paranoid.. shall I remove this?

Reply:
This explanation makes sense to me. I don't think it's necessary to remove this.